client/discovery/discovery.go

```diff
@@ -0,0 +1,133 @@
+	// is used to identify the system in log messages and usually
+	// returns a static string value.
+	Name() string
+
+	// Enabled determines if the discovery layer has been enabled.
+	Enabled() bool
+
+	// DiscoverName returns the name used to register the task into service
+	// discovery. The return value is used in log messages and is passed to
+	// the Register and Deregister functions.
+	DiscoverName(parts []string) string
+
+	// Register is used to register a new entry into a service discovery
+	// system. The name is a symbolic name used to refer to the service. The
+	// context can be used from this function for internal node information
+	// such as IP address. The allocID can be used to uniquely identify the
+	// same task running multiple instances on the same node.
+	Register(allocID, name string, port int) error
+
+	// Deregister is used to deregister a service from a discovery system.
+	Deregister(allocID, name string) error
+}
+
+// Context is used to initialize a discovery backend.
+type Context struct {
+	config *config.Config
+	logger *log.Logger
+	node   *structs.Node
+}
+
+// DiscoveryLayer wraps a set of discovery providers and provides
+// easy calls to register/deregister into all of them.
+type DiscoveryLayer struct {
+	Providers []Provider
+	Context
+}
+
+// NewDiscoveryLayer is used to initialize the discovery layer. It
+// automatically initializes all of the built-in providers and
+// checks their configuration.
+func NewDiscoveryLayer(
+	factories []Factory,
+	config *config.Config,
+	logger *log.Logger,
+	node *structs.Node) (*DiscoveryLayer, error) {
+
+	// Make the context
+	ctx := Context{
+		config: config,
+		logger: logger,
+		node:   node,
+	}
+
+	// Initialize the providers
+	var providers []Provider
+	for _, factory := range factories {
+		provider, err := factory(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed initializing %s discovery: %s",
+				provider.Name(), err)
+		}
+		if provider.Enabled() {
+			providers = append(providers, provider)
+		}
+	}
+	return &DiscoveryLayer{providers, ctx}, nil
+}
+
+// EnabledProviders returns the names of the enabled discovery providers.
+func (d *DiscoveryLayer) EnabledProviders() []string {
+	avail := make([]string, len(d.Providers))
+	for i, provider := range d.Providers {
+		avail[i] = provider.Name()
+	}
+	return avail
+}
+
+// Register iterates all of the providers and registers the given service
+// information with them. If an error is encountered, it is only logged to
+// prevent a single failure from crippling the entire discovery layer.
+func (d *DiscoveryLayer) Register(allocID string, parts []string, port int) {
+	for _, disc := range d.Providers {
+		name := disc.DiscoverName(parts)
+		if err := disc.Register(allocID, name, port); err != nil {
+			d.logger.Printf(
+				"[ERR] client.discovery: error registering %q with %s (alloc %s): %s",
+				parts, disc.Name(), allocID, err)
+			return
```

Comment:
You probably don't want to return here because we potentially could register the service with other providers. It may be worth using multi error and storing all errors to return them. Latter on we could do a retry or some other logic rather than eating the error.